apps/address-service/domains/address/schema/LinkAddressAndSourceService.js

```diff
@@ -0,0 +1,109 @@
+
+const ERRORS = {
+    EMPTY_SOURCE: {
+        query: 'linkAddressAndSource',
+        code: BAD_USER_INPUT,
+        type: EMPTY_SOURCE,
+        message: 'Source is empty',
+    },
+    EMPTY_ADDRESS: {
+        query: 'linkAddressAndSource',
+        code: BAD_USER_INPUT,
+        type: EMPTY_ADDRESS,
+        message: 'Address is empty',
+    },
+    INCORRECT_ADDRESS_SOURCE: {
+        query: 'linkAddressAndSource',
+        code: BAD_USER_INPUT,
+        type: INCORRECT_ADDRESS_SOURCE,
+        message: 'Incorrect address source',
+    },
+    SAME_SOURCE: {
+        mutation: 'createAddressSource',
+        code: BAD_USER_INPUT,
+        type: SOURCE_ALREADY_EXISTS_ERROR,
+        message: 'Source with the same address already exists',
+    },
+}
+
+const LinkAddressAndSourceService = new GQLCustomSchema('LinkAddressAndSourceService', {
+    types: [
+        {
+            access: true,
+            type: 'input LinkAddressAndSourceInput { dv: Int!, sender: JSON!, source: String!, tin: String, address: ID!, parseUnit: Boolean }',
+        },
+        {
+            access: true,
+            type: 'type LinkAddressAndSourceOutput { addressSourceId: ID! }',
+        },
+    ],
+
+    mutations: [
+        {
+            access: access.canLinkAddressAndSource,
+            schema: 'linkAddressAndSource(data: LinkAddressAndSourceInput!): LinkAddressAndSourceOutput',
+            resolver: async (parent, { data }, context) => {
+                let { dv, sender, source, tin, address, parseUnit } = data
+
+                if (!source) {
+                    throw new GQLError({
+                        ...ERRORS.EMPTY_SOURCE,
+                    }, context)
+                }
+
+                if (!address) {
+                    throw new GQLError({
+                        ...ERRORS.EMPTY_ADDRESS,
+                    }, context)
+                }
+
+                if (parseUnit) {
+                    source = parseAddressesFromString([source])[0].result.address
+                }
+
+                if (!source) {
+                    throw new GQLError({
+                        ...ERRORS.INCORRECT_ADDRESS_SOURCE,
+                    }, context)
+                }
+
+                if (tin) {
+                    source = mergeAddressAndHelpers(source, { tin })
+                }
+                try {
+                    const addressSource = await AddressSource.create(
+                        context,
+                        {
+                            dv,
+                            sender,
+                            source,
+                            address: { connect: { id: address } },
+                        },
+                    )
+                    return {
+                        addressSourceId: addressSource.id,
+                    }
+                } catch (e) {
+                    throw new GQLError(ERRORS.SAME_SOURCE, context)
```

Comment:
Is it the only error here?
Maybe it's better to explicitly check for existing the model?